File: trunk/vs2010/Shared/PlateWork/ActivePlate.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using BioNex.Shared.Location;
using BioNex.Shared.PlateDefs;
using BioNex.Shared.TaskListXMLParser;
using log4net;

namespace BioNex.Shared.PlateWork
{
    public abstract class ActivePlate
    {
        // ----------------------------------------------------------------------
        // properties.
        // ----------------------------------------------------------------------
        public static IList< ActivePlate> ActivePlates { get; set; }
        protected static int PlateSerialNumberAutoIncrement { get; set; }

        public int PlateSerialNumber { get; protected set; }

        protected List< PlateTask> ToDoList { get; set; }
        protected IEnumerator< PlateTask> CurrentToDo { get; set; }
        protected bool StillHaveToDos { get; set; }

        public ManualResetEvent PlateIsFree { get; set; }
        public bool Busy{
            get{
                return !PlateIsFree.WaitOne( 0);
            }
        }
        public PlateLocation CurrentLocation { get; set; }
        public PlateLocation DestinationLocation { get; set; }
        public Plate Plate { get; set; }
        public string Barcode{
            get{
                return Plate.Barcode;
            }
        }
        public string LabwareName{
            get{
                return Plate.Labware.Name;
            }
        }
        public int InstanceIndex { get; protected set; }

        public string GetStatus()
        {
            StringBuilder sb = new StringBuilder();
            sb.AppendLine( String.Format( "Info for {0} plate S/N: {1}", this is ActiveSourcePlate ? "source" : "destination", PlateSerialNumber));
            sb.AppendLine( String.Format( "\tInstanceIndex: {0}", InstanceIndex));
            sb.AppendLine( String.Format( "\tBarcode: {0}", Barcode));
            sb.AppendLine( String.Format( "\tLabware: {0}", LabwareName));
            sb.AppendLine( String.Format( "\tBusy: {0}", Busy.ToString()));
            sb.AppendLine( String.Format( "\tCurrent location: {0}{1}{2}{3}", CurrentLocation.Name, CurrentLocation.Available ? ", available" : ", not available", CurrentLocation.Occupied.WaitOne(0) ? ", occupied" : ", not occupied", CurrentLocation.Reserved.WaitOne(0) ? ", reserved" : ", not reserved"));
            sb.AppendLine( String.Format( "\tDestination location: {0}{1}{2}{3}", DestinationLocation.Name, DestinationLocation.Available ? ", available" : ", not available", DestinationLocation.Occupied.WaitOne(0) ? ", occupied" : ", not occupied", CurrentLocation.Reserved.WaitOne(0) ? ", reserved" : ", not reserved"));
            sb.AppendLine( String.Format( "\tPlate is free: {0}", PlateIsFree.WaitOne(0) ? "yes" : "no"));
            sb.AppendLine( "\tToDoList:");
            for( int i=0; i<ToDoList.Count(); i++) {
                PlateTask task = ToDoList[i];
                sb.AppendLine( String.Format( "\t\tTask #{0}", i + 1));
                sb.AppendLine( String.Format( "\t\tDeviceType: {0}", task.DeviceType));
                sb.AppendLine( String.Format( "\t\tCommand: {0}", task.Command));
                sb.AppendLine( String.Format( "\t\tCompleted: {0}", task.Completed));
            }
            return sb.ToString();
        }

        // ----------------------------------------------------------------------
        // members.
        // ----------------------------------------------------------------------
        protected static readonly ILog Log = LogManager.GetLogger( typeof( ActivePlate));

        // ----------------------------------------------------------------------
        // constructors.
        // ----------------------------------------------------------------------
        static ActivePlate()
        {
            ActivePlates = new List< ActivePlate>();
            PlateSerialNumberAutoIncrement = 0;
        }
        // ----------------------------------------------------------------------
        public ActivePlate( Worklist worklist, int instance_index)
        {
            InstanceIndex = instance_index;
            PlateIsFree = new ManualResetEvent( true);
            PlateSerialNumberAutoIncrement++;
            PlateSerialNumber = PlateSerialNumberAutoIncrement;
            Log.InfoFormat( "Created {0}", this);
            // CurrentLocation = new PlateLocation( ToString() + " Home");
            // CurrentLocation.Occupied.Set();
        }

        // ----------------------------------------------------------------------
        // methods.
        // ----------------------------------------------------------------------
        public PlateTask GetCurrentToDo()
        {
            return StillHaveToDos ? CurrentToDo.Current : null;
        }
        // ----------------------------------------------------------------------
        public void AdvanceCurrentToDo()
        {
            StillHaveToDos = CurrentToDo.MoveNext();
        }
        // ----------------------------------------------------------------------
        public virtual bool IsFinished()
        {
            return !Busy && !StillHaveToDos;
        }
        // ----------------------------------------------------------------------
        public void WaitForPlate()
        {
            DestinationLocation.Occupied.WaitOne();
        }
        // ----------------------------------------------------------------------
        public void MarkJobCompleted()
        {
            Log.DebugFormat( "Marking job completed for plate {0}, current location was {1}, destination location was {2}", this, CurrentLocation, DestinationLocation);
            AdvanceCurrentToDo();
            DestinationLocation.Reserved.Reset();
            CurrentLocation = DestinationLocation;
            PlateIsFree.Set();
        }
        // ----------------------------------------------------------------------
        public override string ToString()
        {
            return String.Format( "{0}{1}", GetType().Name, InstanceIndex);
        }
    }

    public class ActiveSourcePlate : ActivePlate
    {
        // ----------------------------------------------------------------------
        // constructors.
        // ----------------------------------------------------------------------
        public ActiveSourcePlate( Worklist worklist, int instance_index)
            : base( worklist, instance_index)
        {
            IList< PlateTask> bb_task_list = new List< PlateTask>{ new PlateTask( "Bumblebee", "source_hitpick")};
            ToDoList = new List< PlateTask>( worklist.TransferOverview.Tasks.source_prehitpick_tasks.Concat( bb_task_list).Concat( worklist.TransferOverview.Tasks.source_posthitpick_tasks));
            CurrentToDo = ToDoList.GetEnumerator();
            StillHaveToDos = CurrentToDo.MoveNext();
            Plate = worklist.SourcePlates[ instance_index];
        }

        // ----------------------------------------------------------------------
        // methods.
        // ----------------------------------------------------------------------
    }

    public class ActiveDestinationPlate : ActivePlate
    {
        // ----------------------------------------------------------------------
        // constructors.
        // ----------------------------------------------------------------------
        public ActiveDestinationPlate( Worklist worklist, int instance_index)
            : base( worklist, instance_index)
        {
            IList< PlateTask> bb_task_list = new List< PlateTask>{ new PlateTask( "Bumblebee", "dest_hitpick")};
            ToDoList = new List< PlateTask>( worklist.TransferOverview.Tasks.dest_prehitpick_tasks.Concat( bb_task_list).Concat( worklist.TransferOverview.Tasks.dest_posthitpick_tasks));
            PlateTask load_task = ToDoList.FirstOrDefault( task => task.DeviceType == "Dock" && task.Command == "Load");
            if( load_task != null){
                string destination_device = worklist.TransferOverview.PlateStorageInterfaceName;
                load_task.ParametersAndVariables.Add( new PlateTask.Parameter( "device_instance", destination_device, ""));
            }
            CurrentToDo = ToDoList.GetEnumerator();
            StillHaveToDos = CurrentToDo.MoveNext();
            Plate = worklist.DestinationPlates[ instance_index];
            // DestinationPlate dst_plate = new DestinationPlate( worklist.TransferOverview.DestinationLabware, instance_index.ToString(), "A1:H12");
            // Plate = dst_plate;
        }

        // ----------------------------------------------------------------------
        // methods.
        // ----------------------------------------------------------------------
    }

    /*
    public class ActiveTipbox : ActivePlate
    {
        // ----------------------------------------------------------------------
        // constructors.
        // ----------------------------------------------------------------------
        public ActiveTipbox( Worklist worklist, int instance_index)
            : base( worklist, instance_index)
        {
            throw new NotImplementedException();
        }

        // ----------------------------------------------------------------------
        // methods.
        // ----------------------------------------------------------------------
    }
    */
}

File: scheduler_framework/scheduler/active_plate.py
"""
ActivePlate - Represents a plate with associated tasks and state.

Based on the C# ActivePlate class, this tracks:
- Plate location (current and destination)
- Task list (ToDo items)
- Plate state (busy/free)
- Completion status
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Optional, Union
from enum import Enum
import threading
from .worklist import PlateTask, WaitTask


class PlateState(Enum):
    """State of a plate in the system"""
    FREE = "free"
    BUSY = "busy"
    MOVING = "moving"
    PROCESSING = "processing"
    COMPLETED = "completed"


@dataclass
class Plate:
    """Represents a physical plate"""
    barcode: str
    labware_name: str
    labware_format: str  # e.g., "96-well", "384-well"
    currently_lidded: bool = True


@dataclass
class PlatePlace:
    """Represents a specific place within a location where a plate can be placed"""
    name: str
    location: 'PlateLocation' = None
    
    def __repr__(self):
        return f"PlatePlace({self.name})"


@dataclass
class PlateLocation:
    """Represents a location where a plate can be placed"""
    name: str
    device_name: str
    available: bool = True
    occupied: threading.Event = field(default_factory=lambda: threading.Event())
    reserved: threading.Event = field(default_factory=lambda: threading.Event())
    places: List[PlatePlace] = field(default_factory=list)
    
    def __post_init__(self):
        # Initially not occupied
        self.occupied.clear()
        self.reserved.clear()
        # Create a default place if none exist
        if not self.places:
            self.places = [PlatePlace(f"{self.name}_place", self)]


class ActivePlate:
    """
    Abstract base class for plates that are actively being processed.
    
    Tracks:
    - Current location and destination
    - Task list (ToDo items)
    - State (busy/free)
    - Completion status
    """
    
    _active_plates: List['ActivePlate'] = []
    _plate_serial_counter = 0
    _lock = threading.Lock()
    
    def __init__(self, worklist: 'Worklist', instance_index: int):
        self.instance_index = instance_index
        self.plate_is_free = threading.Event()
        self.plate_is_free.set()  # Initially free
        
        with ActivePlate._lock:
            ActivePlate._plate_serial_counter += 1
            self.plate_serial_number = ActivePlate._plate_serial_counter
            ActivePlate._active_plates.append(self)
        
        self.current_location: Optional[PlateLocation] = None
        self.destination_location: Optional[PlateLocation] = None
        self.plate: Optional[Plate] = None
        
        # Task management - can contain both PlateTask and WaitTask
        self.todo_list: List[Union[PlateTask, WaitTask]] = []
        self.current_task_index: int = 0
        self.still_have_todos = False
        
    @property
    def busy(self) -> bool:
        """Check if plate is currently busy"""
        return not self.plate_is_free.is_set()
    
    @property
    def barcode(self) -> str:
        """Get plate barcode"""
        return self.plate.barcode if self.plate else ""
    
    @property
    def labware_name(self) -> str:
        """Get labware name"""
        return self.plate.labware_name if self.plate else ""
    
    def get_current_todo(self) -> Optional[Union[PlateTask, WaitTask]]:
        """Get the current task to do (can be PlateTask or WaitTask)"""
        if self.still_have_todos and self.current_task_index < len(self.todo_list):
            return self.todo_list[self.current_task_index]
        return None
    
    def advance_current_todo(self):
        """Move to the next task"""
        self.current_task_index += 1
        self.still_have_todos = self.current_task_index < len(self.todo_list)
    
    def is_finished(self) -> bool:
        """Check if all tasks are completed"""
        return not self.busy and not self.still_have_todos
    
    def mark_job_completed(self):
        """Mark the current job as completed and advance to next task"""
        self.advance_current_todo()
        if self.destination_location:
            self.destination_location.reserved.clear()
        self.current_location = self.destination_location
        self.plate_is_free.set()
    
    def get_status(self) -> str:
        """Get status string for debugging/monitoring"""
        plate_type = "source" if isinstance(self, ActiveSourcePlate) else "destination"
        status = f"Info for {plate_type} plate S/N: {self.plate_serial_number}\n"
        status += f"\tInstanceIndex: {self.instance_index}\n"
        status += f"\tBarcode: {self.barcode}\n"
        status += f"\tLabware: {self.labware_name}\n"
        status += f"\tBusy: {self.busy}\n"
        
        if self.current_location:
            status += f"\tCurrent location: {self.current_location.name}\n"
        if self.destination_location:
            status += f"\tDestination location: {self.destination_location.name}\n"
        
        status += f"\tPlate is free: {self.plate_is_free.is_set()}\n"
        status += "\tToDoList:\n"
        for i, task in enumerate(self.todo_list, 1):
            status += f"\t\tTask #{i}\n"
            status += f"\t\tDeviceType: {task.device_type}\n"
            status += f"\t\tCommand: {task.command}\n"
            status += f"\t\tCompleted: {task.completed}\n"
        
        return status
    
    def __repr__(self):
        return f"{self.__class__.__name__}{self.instance_index}"


class ActiveSourcePlate(ActivePlate):
    """Represents a source plate being processed"""
    
    def __init__(self, worklist: 'Worklist', instance_index: int):
        super().__init__(worklist, instance_index)
        
        # Build task list: pre-hitpick tasks + source_hitpick + post-hitpick tasks
        from .worklist import TransferOverview, PlateTask
        
        if worklist.transfer_overview:
            tasks = []
            # Add pre-hitpick tasks
            tasks.extend(worklist.transfer_overview.tasks.source_prehitpick_tasks)
            # Add the hitpick task
            tasks.append(PlateTask("Bumblebee", "source_hitpick"))
            # Add post-hitpick tasks
            tasks.extend(worklist.transfer_overview.tasks.source_posthitpick_tasks)
            
            self.todo_list = tasks
        
        # Set up task tracking
        self.current_task_index = 0
        self.still_have_todos = len(self.todo_list) > 0
        
        # Get the plate from worklist
        if instance_index < len(worklist.source_plates):
            self.plate = worklist.source_plates[instance_index]


class ActiveDestinationPlate(ActivePlate):
    """Represents a destination plate being processed"""
    
    def __init__(self, worklist: 'Worklist', instance_index: int):
        super().__init__(worklist, instance_index)
        
        # Build task list from transfer overview
        from .worklist import PlateTask
        
        if worklist.transfer_overview:
            tasks = []
            # Add pre-hitpick tasks
            tasks.extend(worklist.transfer_overview.tasks.destination_prehitpick_tasks)
            # Add the hitpick task
            tasks.append(PlateTask("Bumblebee", "destination_hitpick"))
            # Add post-hitpick tasks
            tasks.extend(worklist.transfer_overview.tasks.destination_posthitpick_tasks)
            
            self.todo_list = tasks
        
        # Set up task tracking
        self.current_task_index = 0
        self.still_have_todos = len(self.todo_list) > 0
        
        # Get the plate from worklist
        if instance_index < len(worklist.destination_plates):
            self.plate = worklist.destination_plates[instance_index]

